File: src/main/java/com/sotasan/decompiler/controllers/TabsController.java
package com.sotasan.decompiler.controllers;

import com.sotasan.decompiler.models.FileModel;
import com.sotasan.decompiler.services.ProcessService;
import com.sotasan.decompiler.transformers.Transformer;
import com.sotasan.decompiler.types.ClassType;
import com.sotasan.decompiler.views.TabsView;
import com.sotasan.decompiler.views.TabView;
import lombok.Getter;
import org.jetbrains.annotations.Nullable;
import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

public class TabsController extends BaseController<TabsView> implements ActionListener {

    @Getter private static final TabsController INSTANCE = new TabsController();

    private TabsController() {
        super(new TabsView());
        getView().getComboBox().addActionListener(this);
        String transformer = ProcessService.PREFERENCES.get("transformer", null);
        if (transformer != null)
            getView().getComboBox().setSelectedItem(Transformer.valueOf(transformer));
    }

    @Override
    public void actionPerformed(ActionEvent event) {
        ProcessService.PREFERENCES.put("transformer", getTransformer().name());
        for (int i = 0; i < getView().getTabCount(); i++) {
            TabController controller = ((TabView) getView().getComponentAt(i)).getController();
            if (controller.getFileModel().getType() instanceof ClassType)
                controller.updateAsync();
        }
    }

    public void addTab(FileModel fileModel) {
        TabController controller = getController(fileModel);
        if (controller == null) {
            controller = new TabController(fileModel);
            ImageIcon icon = new ImageIcon(fileModel.getIcon());
            Component component = controller.getComponent();
            controller.updateAsync().thenRun(() -> {
                if (getController(fileModel) == null) {
                    getView().addTab(fileModel.getName(), icon, component);
                    getView().setSelectedComponent(component);
                }
            });
        } else {
            getView().setSelectedComponent(controller.getComponent());
        }
    }

    public void clearTabs() {
        getView().removeAll();
    }

    public Transformer getTransformer() {
        return (Transformer) getView().getComboBox().getSelectedItem();
    }

    private @Nullable TabController getController(FileModel fileModel) {
        TabController controller = null;
        for (int i = 0; i < getView().getTabCount(); i++) {
            TabController current = ((TabView) getView().getComponentAt(i)).getController();
            if (fileModel == current.getFileModel())
                controller = current;
        }
        return controller;
    }

}

File: jvm/src/main/kotlin/moe/sota/decompiler/jvm/controllers/TabsController.kt
package moe.sota.decompiler.jvm.controllers

import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import javax.swing.ImageIcon
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener
import moe.sota.decompiler.jvm.menus.file.FileCloseTab
import moe.sota.decompiler.jvm.models.FileModel
import moe.sota.decompiler.jvm.services.PreferenceService
import moe.sota.decompiler.jvm.transformers.Transformer
import moe.sota.decompiler.jvm.types.ClassType
import moe.sota.decompiler.jvm.views.TabView
import moe.sota.decompiler.jvm.views.TabsView
import org.koin.java.KoinJavaComponent.get

class TabsController(
    private val tabsView: TabsView,
    private val createTabController: (FileModel) -> TabController,
) : ActionListener, ChangeListener {
    val transformer: Transformer?
        get() = tabsView.comboBox.selectedItem as Transformer?

    init {
        tabsView.comboBox.addActionListener(this)
        tabsView.getModel().addChangeListener(this)
        val transformer = PreferenceService.PREFERENCES.get("transformer", null)
        if (transformer != null) tabsView.comboBox.setSelectedItem(Transformer.valueOf(transformer))
    }

    override fun actionPerformed(event: ActionEvent?) {
        PreferenceService.PREFERENCES.put("transformer", this.transformer?.name)
        for (i in 0..<tabsView.tabCount) {
            val controller: TabController? = (tabsView.getComponentAt(i) as TabView).tabController
            if (controller?.fileModel?.type is ClassType) controller.updateAsync()
        }
    }

    override fun stateChanged(changeEvent: ChangeEvent?) {
        val fileCloseTab = get<FileCloseTab>(FileCloseTab::class.java)
        fileCloseTab.setEnabled(tabsView.tabCount > 0)
    }

    fun addTab(fileModel: FileModel) {
        var controller = getController(fileModel)
        if (controller == null) {
            controller = createTabController(fileModel)
            val icon = ImageIcon(fileModel.icon)
            val component = controller.tabView
            controller.updateAsync().thenRun {
                if (getController(fileModel) == null) {
                    tabsView.addTab(fileModel.getName(), icon, component)
                    tabsView.setSelectedComponent(component)
                }
            }
        } else {
            tabsView.setSelectedComponent(controller.tabView)
        }
    }

    fun closeTab() {
        tabsView.removeTabAt(tabsView.selectedIndex)
    }

    fun clearTabs() {
        tabsView.removeAll()
    }

    private fun getController(fileModel: FileModel?): TabController? {
        var controller: TabController? = null

        for (i in 0..<tabsView.tabCount) {
            val current: TabController? = (tabsView.getComponentAt(i) as TabView).tabController
            if (fileModel === current?.fileModel) controller = current
        }

        return controller
    }
}
